blockchain/chain_sync/src/network_context.rs

```diff
@@ -62,6 +64,53 @@ impl<DB: Clone> Clone for SyncNetworkContext<DB> {
     }
 }
 
+/// `RaceBatch` is a structure used to run multiple jobs concurrently
+/// while still being able to limit the amount (`max_concurrent_jobs`).
+/// `get_ok` is the main method that will return the first finishing job with an `Ok` result.
+/// Once at least one job succeed the remaining pending jobs are cancelled.
+/// If none of them succeed then the returned value is `None`.
+struct RaceBatch<T> {
+    tasks: JoinSet<Result<T, String>>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl<T> RaceBatch<T>
+where
+    T: Send + 'static,
+{
+    pub fn new(max_concurrent_jobs: usize) -> Self {
+        RaceBatch {
+            tasks: JoinSet::new(),
+            semaphore: Arc::new(Semaphore::new(max_concurrent_jobs)),
+        }
+    }
+
+    pub fn add(&mut self, future: impl Future<Output = Result<T, String>> + Send + 'static) {
+        let sem = self.semaphore.clone();
+        self.tasks.spawn(async move {
+            let permit = sem
+                .acquire_owned()
+                .await
+                .map_err(|_| "Semaphore unexpectedly closed")?;
+            let result = future.await;
+            drop(permit);
+            result
+        });
+    }
+
+    /// Return first finishing `Ok` future else return `None` if all jobs failed
+    pub async fn get_ok(&mut self) -> Option<T> {
```

Comment:
This should take ownership of `self.` Then `abort_all` won't be necessary (it happens automatically when the `JoinSet` is dropped.)

Reply:
Yes, that's much better.